students/pkoleyni/Lesson03/mailroom.py

```diff
@@ -0,0 +1,162 @@
+        donor_list.append([name])
+    else:
+        print('{} is already in the list of donors'.format(name))
+
+
+def add_donation(name, donation):
+    """
+
+    :param name: name of a donor
+    :param donation: amount of donation
+    :return: add donation to the list and print a message to confirm donation has been added to the list
+    """
+    for i in range(0, len(donor_list)):
+        if donor_list[i][0] == name:
+            donor_list[i].append(donation)
+            print ("${} added to the list of donations made by {} donations".format(donation,name))
+
+
+def thank_you():
+    """
+    This function is taking care of the required action when user select thank you from the menu
+    :return:
+    """
+    usr_input = input('Enter full name of a donor or enter \"List\" for complete list of doors > ')
+    while usr_input == 'List':
+        print_name_of_donors(list_donors)
+        usr_input = input('Enter full name of a donor or enter \"List\" for complete list of doors > ')
+    check_add_name(donor_list, usr_input)
+    usr_donation_input = input('Enter a donation amount > ')
+    add_donation(usr_input, int(usr_donation_input))
+    thank_you_message(usr_input, usr_donation_input)
+
+def thank_you_message(name,donation):
+    """
+    :param name: name of a donot to send thank you letter
+    :param donation: Amount of donation
+    :return:
+    """
+    print()
+    print("Thank you letter for {}".format(name))
+    print('=' * 50)
+    print("Dear {}".format(name))
+    print("We received your {} donation".format(name, donation))
+    print("Thank you for your donation\n")
+
+
+def total_donation(name):
+    """
+
+    :param name: name of a donor
+    :return: will return total donation made by a donor
+    donor_donation[1:] is list of all the donation made by a donor
+    """
+    total = 0
+    for donor_donation in donor_list:
+        if name in donor_donation:
+            for donation in donor_donation[1:]:
+                total = total + donation
+    return total
+
+def num_gifts(name):
+    """
+
+    :param name: is name of a donor
+    :return: will return total number of gifts from a donor
+    """
+    for donor_donation in donor_list:
+        if name in donor_donation:
+            total_num_gifts = len(donor_donation[1:])
+    return total_num_gifts
+
+def average_gifts(name):
+    for donor_donation in donor_list:
+        if name in donor_donation:
+            avarage = sum(donor_donation[1:]) / len(donor_donation[1:])
+    return avarage
+
+def report():
+    """
+
+    :return: will call other function to get the data and print the report
+    """
+    print()
+    report_head = ("{:<20}{:<20}{:<20}{:<20}".format('Donor Name','| Total Given', '| Num Gifts ', '| Average Gift' ))
+    print (report_head)
+    print ('-'*(len(report_head)))
+    for donor_name in list_donors():
+        print(f"{donor_name:<20}{total_donation(donor_name):<20f}{num_gifts(donor_name):<20f}{average_gifts(donor_name)}")
```

Comment:
missing sort feature

Reply:
also if you're supporting fractional donations, you should only do two decimal points.

Reply:
you're looping 3 additional times in total_donation(), num_gifts(), and average_gifts() it's wasteful - and such simple code that you can do all in one loop.

```
for donor in donations:
    donations = donor[1:]
    total = sum(donations)
    num_of_gifts = len(donations)
    avg_gifts = total / num_of_gifts
```

Reply:
Then if i am not wrong i will need to have one function to return 3 items, i thought it is better to have functions which are doing one specific task.
Can you please explain which one is better ? multiple functions or one function to loop once and return 3 items

Reply:
You don't need any functions in this particular case - take a look at my example.

Python is big on simplicity, the code is so simple that adding functions just adds unnecessary complexity _and_ inefficiency. I'm not saying not to ever use functions, just not in this particular case.